src/Lecture2.hs

```diff
@@ -288,4 +326,13 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding = go 0
+  where
+    go :: Int -> Expr -> Expr
+    go acc (Add (Lit i1) (Lit i2)) =  Lit (acc + i1 + i2)
+    go acc (Add (Lit i) e2) = go (acc + i) e2
+    go acc (Add e1 (Lit i)) =  go (acc + i) e1
+    go acc e
+        | 0 == acc =  e
+        | otherwise  =  Add (Lit acc) e
```

Comment:
The problem with this case is that it doesn't go deeper inside if you have `Add` of two nested `Add`s inside. The current implementation won't fold constants deep inside